core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -138,8 +139,8 @@ class LogSegment(val log: FileMessageSet,
     val length = 
       maxOffset match {
         case None =>
-          // no max offset, just use the max size they gave unmolested
-          maxSize
+          // no max offset, just read until the max position
+          (maxPosition.getOrElse(size) - startPosition.position).toInt
```

Comment:
I don't think so. Although the type is Long, but both numbers are actually positive integer, otherwise we would have issue for log index because it uses integer as position. So the subtraction of two integers won't cause overflow. 

I don't think the result would be negative either, because the starting offset should never be higher than the last exposed MessageSet's highest offset. And the maxPosition will always be greater or equal to the last exposed MessageSet's highest offset.